kolibri/auth/models.py

```diff
@@ -1,145 +1,293 @@
+            raise MorangoError(
+                "Cannot filter model '{model}' on nonexistent partitions '{partitions}'.".format(
+                    model=self.model.__name__,
+                    partitions=nonexistent_partitions
+                )
+            )
+
+        # for each of the provided arguments, add a filter query parameter mapped to the appropriate model field
+        for key, val in kwargs.items():
+
+            field_name = self.model._morango_partition_fields[key]
+
+            # if attempting to filter on a partition for which the current model has no field, there are no matches
+            if not hasattr(self.model, field_name):
+                return self.none()
+
+            query[field_name] = val
+
+        # apply the calculated filter to the queryset
+        return self.filter(**query)
+
+
+class SyncableModel(models.Model):
+
+    _morango_partition_fields = {}  # maps from a partition name to a model field name
+
+    objects = SyncableModelPartitionQuerySet.as_manager()
+    # special reference to partition manager in case 'objects' is overridden in subclasses:
+    partitioned_objects = SyncableModelPartitionQuerySet.as_manager()
+
+    id = UUIDField(primary_key=True)
+
+    class Meta:
+        abstract = True
+
+    def calculate_uuid(self):
+        return uuid.uuid4().hex  # for now, in this dummy stub, we're just using random UUIDs
+
+    def get_partition(self):
+        """
+        Calculate and return the partition attributes for the current model instance.
+        """
+        partition = {}
+        for partition_name, field_name in self._morango_partition_fields.items():
+            if hasattr(self, field_name):
+                partition[partition_name] = getattr(self, field_name)
+        return partition
+
+    def clean(self):
+        if not self.id:
+            self.id = self.calculate_uuid()
+        super(SyncableModel, self).clean()
+
+
+class MorangoTreeManager(TreeManager):
+
+    # Override the logic in Django MPTT that assigns tree ids, as we need to ensure tree ids do not
+    # conflict for facilities created on differentdevices and then later synced to the same device
+    # By default, tree ids are auto-increasing integers, but we use UUIDs to avoid collisions.
+    def _get_next_tree_id(self):
+        return uuid.uuid4().hex
+
+
+class MorangoMPTTModel(MPTTModel):
+    """
+    Any model that inherits from SyncableModel that wants to inherit from MPTTModel should instead inherit
+    from MorangoMPTTModel, which modifies some behavior to make it safe for the syncing system.
+    """
+
+    _default_manager = MorangoTreeManager()
+
+    # change tree_id to a uuid to avoid collisions; see explanation above in the MorangoTreeManager class
+    tree_id = UUIDField()
+
+    class Meta:
+        abstract = True
+
+
+##############################################
+# /END MORANGO
+##############################################
+
+
+class KolibriSyncableModel(SyncableModel):
```

Comment:
This whole class, for instance, is tightly coupled to Morango. If this is to be the base class for all or most of Kolibri's Models, should we not scrutinize it now at least to the extent of agreeing that it provides a reasonable API?